backend.native/tests/runtime/basic/cleaner_basic.kt

```diff
@@ -0,0 +1,188 @@
+import kotlin.native.concurrent.*
+import kotlin.native.ref.WeakReference
+import kotlin.time.*
+
+class AtomicBoolean(initialValue: Boolean) {
+    private val impl = AtomicInt(if (initialValue) 1 else 0)
+
+    init {
+        freeze()
+    }
+
+    public var value: Boolean
+        get() = impl.value != 0
+        set(new) { impl.value = if (new) 1 else 0 }
+}
+
+class FunBox(private val impl: () -> Unit) {
+    fun call() {
+        impl()
+    }
+}
+
+@Test
+fun testCleaner() {
+    val called = AtomicBoolean(false);
+    var funBoxWeak: WeakReference<FunBox>? = null
+    var cleanerWeak: WeakReference<Cleaner>? = null
+    {
+        val cleaner = {
+            val funBox = FunBox { called.value = true }.freeze()
+            funBoxWeak = WeakReference(funBox)
+            createCleaner(funBox) { it.call() }
+        }()
+        GC.collect()  // Make sure local funBox reference is gone
+        cleanerWeak = WeakReference(cleaner)
+        assertFalse(called.value)
+    }()
+
+    GC.collect()
+    scheduleGCOnCleanerWorker().result
+
+    assertNull(cleanerWeak!!.value)
+    assertTrue(called.value)
+    assertNull(funBoxWeak!!.value)
+}
+
+@Test
+fun testCleanerFrozen() {
+    val called = AtomicBoolean(false);
+    var funBoxWeak: WeakReference<FunBox>? = null
+    var cleanerWeak: WeakReference<Cleaner>? = null
+    {
+        val cleaner = {
+            val funBox = FunBox { called.value = true }.freeze()
+            funBoxWeak = WeakReference(funBox)
+            createCleaner(funBox) { it.call() }
+        }()
+        GC.collect()  // Make sure local funBox reference is gone
+        cleaner.freeze()
+        cleanerWeak = WeakReference(cleaner)
+        assertFalse(called.value)
+    }()
+
+    GC.collect()
+    scheduleGCOnCleanerWorker().result
+
+    assertNull(cleanerWeak!!.value)
+    assertTrue(called.value)
+    assertNull(funBoxWeak!!.value)
+}
+
+inline fun tryWithTimeout(timeoutSeconds: Int, f: () -> Unit): Unit {
+    val timeout = TimeSource.Monotonic.markNow() + timeoutSeconds.seconds
+    while (true) {
+        try {
+            f()
+            return
+        } catch (e: Throwable) {
+            if (timeout.hasPassedNow()) {
+                throw e
+            }
+        }
+    }
+}
+
+@Test
+fun testCleanerCleansWithoutGC() {
+    val called = AtomicBoolean(false);
+    var funBoxWeak: WeakReference<FunBox>? = null
+    var cleanerWeak: WeakReference<Cleaner>? = null
+    {
+        val cleaner = {
+            val funBox = FunBox { called.value = true }.freeze()
+            funBoxWeak = WeakReference(funBox)
+            createCleaner(funBox) { it.call() }
+        }()
+        GC.collect()  // Make sure local funBox reference is gone
+        cleaner.freeze()
+        cleanerWeak = WeakReference(cleaner)
+        assertFalse(called.value)
+    }()
+
+    GC.collect()
+
+    assertNull(cleanerWeak!!.value)
+    tryWithTimeout(3) {
+        GC.collect()  // Collect local stack (from previous iteration)
```

Comment:
Enclosing lambda gets inlined, so the code in the block can actually retain some strong references to the objects. Is this intended?

Reply:
It gets inlined, yes, and the resulting code should look like
```kotlin
while (true) {
    try {
        GC.collect()
        // retain on stack and possibly throw an exception
        return
    } catch(…) { … }
}
```
So, if the exception happens, the newly retained objects will be out of scope by the next `while` iteration and `GC.collect()` should collect them.

Reply:
I see. Ok.